File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
    min=0
    max=0
    profit=0
    for i in 1..prices.length-1
        if prices[i]>prices[max]
            max=i
            new_profit=prices[max]-prices[min]
           profit=new_profit if new_profit>profit
         elsif prices[i]<prices[min]
             min=i
             max=i
        end
    end
    profit
end

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.ts
function maxProfit(prices: number[]): number {
    /*
    * profit=0
    * left=0
    * right=1
    * while(right<nums.length)
    * if(profit<right-left)
    * set new profit
    * right<left
    * left=right
    * return profit
    */

let profit = 0;
let left=0;
let right=1
    while(right<prices.length){
        if(profit<(prices[right]-prices[left]))
            profit = prices[right]-prices[left];
        else if(prices[right]<prices[left])
            left=right;
        right++
    }
    return profit
};